File: 백준/Silver/17300. 패턴/패턴.py
import sys
input = sys.stdin.readline

def No():
    print("NO")
    quit()

l = int(input())
visited = [0] * 10
a = list(map(int, input().split()))

for j in range(l):
    prev = a[j-1]
    i = a[j]
    if visited[i] == 1:
        No()
    visited[i] = 1
    if j == 0:
        continue

    if visited[2] == 0:
        if (prev+i) == 4:
            No()
    if visited[4] == 0:
        if (prev == 1 and i == 7) | (prev == 7 and i == 1):
            No()
    if visited[5] == 0:
        if (prev+i) == 10:
            No()
    if visited[6] == 0:
        if (prev == 3 and i == 9) | (prev == 9 and i == 3):
            No()
    if visited[8] == 0:
        if (prev+i) == 16:
            No()

print("YES")

File: 백준/Silver/17300. 패턴/패턴.js
const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./example.txt")
  .toString()
  .trim()
  .split("\n")
  .map((v) => v.split(" ").map(Number));

const [l, a] = input;
const visited = [...new Array(10)].fill(false);
visited[a[0]] = true;

function pattern(v) {
  for (let i = 1; i < l; i++) {
    let prev = a[i - 1];
    let now = a[i];
    if (v[now]) return "NO";
    v[now] = true;

    if (!v[2]) {
      if (prev + now === 4) return "NO";
    }
    if (!v[4]) {
      if ((prev === 1 && now === 7) || (prev === 7 && now === 1)) return "NO";
    }
    if (!v[5]) {
      if (prev + now === 10) return "NO";
    }
    if (!v[6]) {
      if ((prev === 3 && now === 9) || (prev === 9 && now === 3)) return "NO";
    }
    if (!v[8]) {
      if (prev + now === 16) return "NO";
    }
  }
  return "YES";
}
console.log(pattern(visited));
